File: 777.在LR字符串中交换相邻字符/main.ts
function canTransform(start: string, end: string): boolean {
  const sLen = start.length;
  let i = 0;
  let j = 0;

  while (i < sLen && j < sLen) {
    // 遍历start和end非"X"的字符
    while (i < sLen && start[i] === "X") {
      i++;
    }
    while (j < sLen && end[j] === "X") {
      j++;
    }

    // 如果不等，一定返回false
    if (start[i] !== end[j]) {
      return false;
    }

    // 如果相等，判断当前字符
    // 为L，且在start中较早出现，则一定返回false
    // 为R，且在end中较早出现，则一定返回false
    const c = start[i];
    if ((c === "L" && i < j) || (c === "R" && i > j)) {
      return false;
    }
    i++;
    j++;
  }

  // 当end遍历完，start中还有非X字符一定返回false
  while (i < sLen) {
    if (start[i] !== "X") {
      return false;
    }
    i++;
  }

  // 当start遍历完，end中还有非X字符一定返回false
  while (j < sLen) {
    if (end[j] !== "X") {
      return false;
    }
    j++;
  }
  return true;
}

File: 777.在LR字符串中交换相邻字符/main.rs
pub fn can_transform(start: String, end: String) -> bool {
    start.chars().filter(|&c| c == 'X').count() == end.chars().filter(|&c| c == 'X').count()
        && start
            .chars()
            .enumerate()
            .filter(|c| c.1 != 'X')
            .zip(end.chars().enumerate().filter(|c| c.1 != 'X'))
            .all(|(c1, c2)| match (c1.1, c2.1) {
                ('L', 'L') => c1.0 >= c2.0,
                ('R', 'R') => c1.0 <= c2.0,
                _ => false,
            })
}
